lib/puppet/resource_api/parameter.rb

```diff
@@ -0,0 +1,48 @@
+require 'puppet/parameter'
+
+module Puppet; module ResourceApi; end; end # predeclare the main module # rubocop:disable Style/Documentation,Style/ClassAndModuleChildren
+
+# Class containing parameter functionality for ResourceApi.
+class Puppet::ResourceApi::Parameter < Puppet::Parameter
+  # This initialize takes arguments and sets up new parameter.
+  # @param name the name used for reporting errors
+  # @param type the type instance
+  # @param definition the definition of the property
+  # @param resource the resource instance which is passed to the parent class.
+  def initialize(name, type, definiton, resource)
```

Comment:
To clean this up in relation to my comment above around the `data_type` function, we need to have a good look at how this interface looks and works to ensure that we do not just move from one maintenance head-ache to the other.

Looking at coupling and cohesion, the signature as it stands has problems:

* too high coupling to `resource_api.rb` through passing the full `definition`.
* not enough cohesion with core puppet by requiring unpacking and repacking of the `resource` parameter.
* using `type` instead of `data_type` - thanks to the overloading of terms, we **always** have to be very specific of which kind of type (data vs resource type) we mean
* lack of extensibility/future-proofing, by not passing in the `options` of this attribute

To address these issues, a better signature could look like this:

```ruby
def initialize(type_name, data_type, attribute_name, resource_hash)
```

* provides the immediately necessary data for the object with no additional processing required 
* lists parameters in an order similar to a declaration in typed languages (loosely modelled after something like `class Foo { int bar; }`, with `type_name` being `Foo`, `data_type` being `int`, and `attribute_name` being `bar` )
* re-use puppet core's `resource_hash`, which gets passed through to `super` unmodified

Looking back at `resource_api.rb`, usage would now look like this:

```ruby
define_method(:initialize) do |resource_hash|
  super(definition[:name], self.class.data_type, name, resource_hash)
```

PS: typo: `definiton` -> `definition`
PPS: it would be helpful to rename `name` to `attribute_name` in the entire loop in `resource_api.rb`.

Reply:
I will change it, it looks like much better choice that what it is now